modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/CollectPartitionKeysByRecheckRequestTask.java

```diff
@@ -0,0 +1,207 @@
+
+            ExecutionResult<List<PartitionDataRow>> excRes = result.getData();
+
+            if (excRes.getErrorMessage() != null)
+                return new ExecutionResult<>(excRes.getErrorMessage());
+
+            List<PartitionDataRow> partKeys = excRes.getResult();
+
+            for (PartitionDataRow key : partKeys) {
+                try {
+                    KeyCacheObject keyObj = unmarshalKey(key.getKey(), ctx);
+                    res.computeIfAbsent(keyObj, k -> new HashMap<>()).put(
+                        key.getNodeId(),
+                        new VersionedValue(key.getVal(), key.getVersion(), key.getUpdateCounter(), key.getRecheckStartTime())
+                    );
+                }
+                catch (Exception e) {
+                    U.error(log, e.getMessage(), e);
+
+                    return new ExecutionResult<>(e.getMessage());
+                }
+            }
+        }
+
+        return new ExecutionResult<>(res);
+    }
+
+    /**
+     *
+     */
+    public static class CollectRecheckJob extends ReconciliationResourceLimitedJob {
+        /**
+         *
+         */
+        private static final long serialVersionUID = 0L;
+
+        /** Partition key. */
+        private final RecheckRequest recheckReq;
+
+        /**
+         * @param recheckReq Recheck request.
+         */
+        public CollectRecheckJob(RecheckRequest recheckReq) {
+            this.recheckReq = recheckReq;
+        }
+
+        /** {@inheritDoc} */
+        @Override protected long sessionId() {
+            return recheckReq.sessionId();
+        }
+
+        /** {@inheritDoc} */
+        @Override protected ExecutionResult<List<PartitionDataRow>> execute0() {
+            GridCacheContext<Object, Object> cctx = ignite.context().cache().cache(recheckReq.cacheName()).context();
+
+            CacheGroupContext grpCtx = cctx.group();
+
+            GridDhtLocalPartition part = grpCtx.topology().localPartition(recheckReq.partitionId());
+
+            assert part != null;
+
+            part.reserve();
+
+            List<PartitionDataRow> recheckedKeys = new ArrayList<>();
+
+            long updateCntr = part.updateCounter();
+            long recheckStartTime = System.currentTimeMillis();
+
+            try {
+                for (KeyCacheObject recheckKey : recheckReq.recheckKeys()) {
+                    try {
+                        KeyCacheObject key = unmarshalKey(recheckKey, cctx);
+
+                        CacheDataRow row = grpCtx.offheap().dataStore(part).find(cctx, key);
+
+                        if (row != null)
+                            recheckedKeys.add(new PartitionDataRow(ignite.localNode().id(), row.key(), row.version(), row.value(), updateCntr, recheckStartTime));
+                    }
+                    catch (IgniteCheckedException e) {
+                        String errMsg = "Recheck key [" + recheckKey + "] was skipped.";
```

Comment:
Inappropriate log message format. Correct something like this: `Recheck key skipped [key=" + recheckKey + ']' `